[libunwind] Tweak tests for musl support.

al45tair opened the pull request:
We can't use `dladdr()` in the tests, because when we're statically linking with musl that function is a no-op.

Additionally, because musl disables emission of unwind information in its build, and because its signal trampolines don't include unwind information, tests that involve unwinding through a signal handler won't work and need to be disabled for musl.

rdar://123436891

al45tair committed:
[libunwind] Tweak tests for musl support.

We can't use `dladdr()` in the tests, because when we're statically
linking with musl that function is a no-op.

Additionally, because musl disables emission of unwind information
in its build, and because its signal trampolines don't include
unwind information, tests that involve unwinding through a signal
handler won't work and need to be disabled for musl.

rdar://123436891

al45tair commented:
This is a cherry pick of https://github.com/apple/llvm-project/pull/8260.

al45tair committed:
[libunwind] Tweak formatting slightly.

Tweak code formatting to keep clang-format happy.

rdar://123436891

arichardson commented on libunwind/test/floatregister.pass.cpp:
```diff
@@ -11,20 +11,27 @@
 #include <string.h>
 #include <unwind.h>
 
+// Using __attribute__((section("main_func"))) is Linux specific, but then
```
Nit: This is not specific to Linux it is specific to the behaviour of ELF linkers that automatically create these symbols for sections that are valid C identifiers.

al45tair replied:
Fair point.

al45tair commented on libunwind/test/signal_unwind.pass.cpp:
```diff
@@ -24,16 +23,30 @@
 #include <unistd.h>
 #include <unwind.h>
 
+// Note: this test fails on musl because:
+//
+//  (a) musl disables emission of unwind information for its build, and
+//  (b) musl's signal trampolines don't include unwind information
+//
```
Let's rearrange it to use `XFAIL`, since that's clearer.

al45tair commented on libunwind/test/unwind_leaffunction.pass.cpp:
```diff
@@ -24,16 +23,30 @@
 #include <unistd.h>
 #include <unwind.h>
 
+// Note: this test fails on musl because:
+//
+//  (a) musl disables emission of unwind information for its build, and
+//  (b) musl's signal trampolines don't include unwind information
+//
```
We'll use `XFAIL` here too.

al45tair committed:
[libunwind] Fix nits from review.

Fix a couple of comments, and change to using `XFAIL` rather than explicitly
selecting only `linux-gnu`.

rdar://123436891

al45tair commented:
I don't have merge rights here, so someone else will need to hit the Merge button.
